File: packages/domain/src/query-packs/registry.ts
/**
 * In-memory query pack registry for versioned pack lookup (BB-038).
 */
import { assertQueryPackValid } from './pack.js';
import type { EntityKind } from '../entity-kinds.js';
import type { QueryPack, QueryPackTheme } from './types.js';

export type QueryPackRegistryStore = {
  readonly packs: Map<string, QueryPack>;
};

export function createInMemoryQueryPackRegistry(): QueryPackRegistryStore {
  return { packs: new Map() };
}

export function registerQueryPack(store: QueryPackRegistryStore, pack: QueryPack): QueryPack {
  assertQueryPackValid(pack);
  store.packs.set(pack.id, pack);
  return pack;
}

export function getQueryPack(store: QueryPackRegistryStore, id: string): QueryPack | undefined {
  return store.packs.get(id);
}

export type ListQueryPacksFilter = {
  readonly entityKind?: EntityKind;
  readonly theme?: QueryPackTheme;
};

export function listQueryPacks(
  store: QueryPackRegistryStore,
  filter: ListQueryPacksFilter = {},
): readonly QueryPack[] {
  return [...store.packs.values()].filter((pack) => {
    if (filter.entityKind !== undefined && pack.entityKind !== filter.entityKind) {
      return false;
    }
    if (filter.theme !== undefined && pack.theme !== filter.theme) {
      return false;
    }
    return true;
  });
}

export function resolveQueryPackForRun(
  store: QueryPackRegistryStore,
  input: { readonly entityKind: EntityKind; readonly theme: QueryPackTheme },
): QueryPack {
  const matches = listQueryPacks(store, input);
  if (matches.length === 0) {
    throw new Error(
      `No query pack registered for entityKind=${input.entityKind} theme=${input.theme}`,
    );
  }
  if (matches.length > 1) {
    throw new Error(
      `Ambiguous query pack for entityKind=${input.entityKind} theme=${input.theme}`,
    );
  }
  return matches[0]!;
}

File: workers/research/src/black_book_research/query_packs/registry.py
"""In-memory query pack registry (BB-038)."""

from __future__ import annotations

from dataclasses import dataclass, field

from .pack import assert_query_pack_valid
from .types import EntityKind, QueryPack, QueryPackTheme


@dataclass
class QueryPackRegistryStore:
    packs: dict[str, QueryPack] = field(default_factory=dict)


def create_in_memory_query_pack_registry() -> QueryPackRegistryStore:
    return QueryPackRegistryStore()


def register_query_pack(store: QueryPackRegistryStore, pack: QueryPack) -> QueryPack:
    assert_query_pack_valid(pack)
    store.packs[pack.id] = pack
    return pack


def get_query_pack(store: QueryPackRegistryStore, pack_id: str) -> QueryPack | None:
    return store.packs.get(pack_id)


def list_query_packs(
    store: QueryPackRegistryStore,
    *,
    entity_kind: EntityKind | None = None,
    theme: QueryPackTheme | None = None,
) -> tuple[QueryPack, ...]:
    results: list[QueryPack] = []
    for pack in store.packs.values():
        if entity_kind is not None and pack.entity_kind != entity_kind:
            continue
        if theme is not None and pack.theme != theme:
            continue
        results.append(pack)
    return tuple(results)


def resolve_query_pack_for_run(
    store: QueryPackRegistryStore,
    *,
    entity_kind: EntityKind,
    theme: QueryPackTheme,
) -> QueryPack:
    matches = list_query_packs(store, entity_kind=entity_kind, theme=theme)
    if not matches:
        raise ValueError(f"No query pack registered for entity_kind={entity_kind} theme={theme}")
    if len(matches) > 1:
        raise ValueError(f"Ambiguous query pack for entity_kind={entity_kind} theme={theme}")
    return matches[0]
